File: frontend_v2/src/components/CategoryManagement.js
import React, { useState, useEffect } from 'react';
import { FaFolder, FaPlus, FaEdit, FaTrash } from 'react-icons/fa';
import apiClient from '../utils/apiClient';

const CategoryManagement = () => {
  const [categories, setCategories] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  const [activeTab, setActiveTab] = useState('view');
  const [formData, setFormData] = useState({
    name: '',
    description: ''
  });
  const [selectedCategory, setSelectedCategory] = useState(null);
  const [successMessage, setSuccessMessage] = useState('');

  useEffect(() => {
    fetchCategories();
  }, []);

  const fetchCategories = async () => {
    try {
      setLoading(true);
      const data = await apiClient.getCategories();
      setCategories(data || []);
    } catch (err) {
      setError('Failed to load categories');
      console.error('Categories error:', err);
    } finally {
      setLoading(false);
    }
  };

  const handleCreateCategory = async (e) => {
    e.preventDefault();
    if (!formData.name.trim()) {
      setError('Category name is required!');
      return;
    }

    try {
      await apiClient.createCategory(formData.name.trim());
      setSuccessMessage(`✅ Category '${formData.name}' created successfully!`);
      setFormData({ name: '', description: '' });
      fetchCategories();
    } catch (err) {
      setError('❌ Failed to create category. Please try again.');
    }
  };

  const handleUpdateCategory = async (e) => {
    e.preventDefault();
    if (!selectedCategory) return;

    try {
      await apiClient.updateCategory(
        selectedCategory.id,
        formData.name.trim(),
        formData.description.trim() || null
      );
      setSuccessMessage('✅ Category updated successfully!');
      setSelectedCategory(null);
      setFormData({ name: '', description: '' });
      fetchCategories();
    } catch (err) {
      setError('❌ Failed to update category. Please try again.');
    }
  };

  const handleDeleteCategory = async (categoryId) => {
    if (!window.confirm('Are you sure you want to delete this category?')) return;

    try {
      await apiClient.deleteCategory(categoryId);
      setSuccessMessage('✅ Category deleted successfully!');
      fetchCategories();
    } catch (err) {
      setError('❌ Failed to delete category. Make sure it has no items.');
    }
  };

  const handleEditCategory = (category) => {
    setSelectedCategory(category);
    setFormData({
      name: category.name,
      description: category.description || ''
    });
    setActiveTab('edit');
  };

  const clearMessages = () => {
    setError('');
    setSuccessMessage('');
  };

  if (loading) {
    return (
      <div style={{ textAlign: 'center', padding: '50px' }}>
        <div>Loading categories...</div>
      </div>
    );
  }

  return (
    <div>
      <div style={{ marginBottom: '30px' }}>
        <h1 style={{ display: 'flex', alignItems: 'center', marginBottom: '10px' }}>
          <FaFolder style={{ marginRight: '10px', color: '#007bff' }} />
          📂 Category Management
        </h1>
      </div>

      {/* Tabs */}
      <div className="tabs">
        <button
          className={`tab ${activeTab === 'view' ? 'active' : ''}`}
          onClick={() => { setActiveTab('view'); clearMessages(); }}
        >
          📋 View Categories
        </button>
        <button
          className={`tab ${activeTab === 'create' ? 'active' : ''}`}
          onClick={() => { setActiveTab('create'); clearMessages(); }}
        >
          ➕ Create Category
        </button>
        <button
          className={`tab ${activeTab === 'edit' ? 'active' : ''}`}
          onClick={() => { setActiveTab('edit'); clearMessages(); }}
        >
          ✏️ Edit Categories
        </button>
      </div>

      {/* Messages */}
      {error && (
        <div className="alert alert-danger">
          {error}
        </div>
      )}
      {successMessage && (
        <div className="alert alert-success">
          {successMessage}
        </div>
      )}

      {/* View Categories Tab */}
      <div className={`tab-content ${activeTab === 'view' ? 'active' : ''}`}>
        <div className="card">
          <h3>All Categories</h3>
          {categories.length > 0 ? (
            categories.map((category) => (
              <div key={category.id} className="expander">
                <div className="expander-header">
                  <span>📁 {category.name}</span>
                  <button
                    className="btn btn-danger"
                    onClick={() => handleDeleteCategory(category.id)}
                    style={{ padding: '5px 10px', fontSize: '12px' }}
                  >
                    <FaTrash style={{ marginRight: '5px' }} />
                    Delete
                  </button>
                </div>
                <div className="expander-content">
                  <div className="grid grid-3">
                    <div>
                      <strong>Name:</strong> {category.name}
                    </div>
                    <div>
                      <strong>ID:</strong> {category.id}
                    </div>
                    <div>
                      <strong>Description:</strong> {category.description || 'No description'}
                    </div>
                  </div>
                </div>
              </div>
            ))
          ) : (
            <div className="alert alert-info">
              No categories found. Create your first category in the 'Create Category' tab!
            </div>
          )}
        </div>
      </div>

      {/* Create Category Tab */}
      <div className={`tab-content ${activeTab === 'create' ? 'active' : ''}`}>
        <div className="card">
          <h3>Create New Category</h3>
          <form onSubmit={handleCreateCategory}>
            <div className="form-group">
              <label>Category Name</label>
              <input
                type="text"
                className="form-control"
                value={formData.name}
                onChange={(e) => setFormData({ ...formData, name: e.target.value })}
                placeholder="e.g., Beverages, Dairy, Snacks"
                required
              />
            </div>
            <div className="form-group">
              <label>Description (optional)</label>
              <textarea
                className="form-control"
                value={formData.description}
                onChange={(e) => setFormData({ ...formData, description: e.target.value })}
                placeholder="Describe what items belong in this category"
                rows="3"
              />
            </div>
            <button type="submit" className="btn btn-primary">
              Create Category
            </button>
          </form>
        </div>
      </div>

      {/* Edit Categories Tab */}
      <div className={`tab-content ${activeTab === 'edit' ? 'active' : ''}`}>
        <div className="card">
          <h3>Edit Categories</h3>
          {categories.length > 0 ? (
            <div>
              <div className="form-group">
                <label>Select Category to Edit:</label>
                <select
                  className="form-control"
                  value={selectedCategory?.id || ''}
                  onChange={(e) => {
                    const category = categories.find(c => c.id === parseInt(e.target.value));
                    if (category) {
                      handleEditCategory(category);
                    }
                  }}
                >
                  <option value="">Select a category...</option>
                  {categories.map((category) => (
                    <option key={category.id} value={category.id}>
                      {category.name}
                    </option>
                  ))}
                </select>
              </div>

              {selectedCategory && (
                <form onSubmit={handleUpdateCategory}>
                  <div className="form-group">
                    <label>New Name</label>
                    <input
                      type="text"
                      className="form-control"
                      value={formData.name}
                      onChange={(e) => setFormData({ ...formData, name: e.target.value })}
                      required
                    />
                  </div>
                  <div className="form-group">
                    <label>New Description</label>
                    <textarea
                      className="form-control"
                      value={formData.description}
                      onChange={(e) => setFormData({ ...formData, description: e.target.value })}
                      rows="3"
                    />
                  </div>
                  <div className="grid grid-2">
                    <button type="submit" className="btn btn-primary">
                      Update Category
                    </button>
                    <button
                      type="button"
                      className="btn btn-danger"
                      onClick={() => handleDeleteCategory(selectedCategory.id)}
                    >
                      Delete Category
                    </button>
                  </div>
                </form>
              )}
            </div>
          ) : (
            <div className="alert alert-info">
              No categories to edit. Create some categories first!
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default CategoryManagement; 

File: frontend/pages/category_management.py
import streamlit as st
from utils.api_client import api_client
from utils.auth import check_auth, show_logout_button

# Check authentication first
check_auth()

# Show logout button in sidebar
show_logout_button()

st.set_page_config(
    page_title="Category Management - Stocker",
    page_icon="📂",
    layout="wide"
)

st.title("📂 Category Management")

# Get existing categories
categories = api_client.get_categories()

# Create tabs for different operations
tab1, tab2, tab3 = st.tabs(["📋 View Categories", "➕ Create Category", "✏️ Edit Categories"])

with tab1:
    st.subheader("All Categories")
    
    if categories:
        # Display categories in a table format
        for category in categories:
            with st.expander(f"📁 {category.get('name', 'Unknown')}"):
                col1, col2, col3 = st.columns([2, 1, 1])
                
                with col1:
                    st.write(f"**Name:** {category.get('name', 'Unknown')}")
                    if category.get('description'):
                        st.write(f"**Description:** {category.get('description', 'No description')}")
                    else:
                        st.write("**Description:** No description")
                
                with col2:
                    st.write(f"**ID:** {category.get('id', 'Unknown')}")
                
                with col3:
                    # Delete button
                    if st.button("🗑️ Delete", key=f"delete_{category.get('id')}"):
                        result = api_client.delete_category(category.get('id'))
                        if result:
                            st.success("✅ Category deleted successfully!")
                            st.rerun()
                        else:
                            st.error("❌ Failed to delete category. Make sure it has no items.")
    else:
        st.info("No categories found. Create your first category in the 'Create Category' tab!")

with tab2:
    st.subheader("Create New Category")
    
    with st.form("create_category_form"):
        category_name = st.text_input("Category Name", placeholder="e.g., Beverages, Dairy, Snacks")
        category_description = st.text_area("Description (optional)", placeholder="Describe what items belong in this category")
        
        if st.form_submit_button("Create Category"):
            if category_name.strip():
                result = api_client.create_category(category_name.strip())
                if result:
                    st.success(f"✅ Category '{category_name}' created successfully!")
                    st.rerun()
                else:
                    st.error("❌ Failed to create category. Please try again.")
            else:
                st.error("❌ Category name is required!")

with tab3:
    st.subheader("Edit Categories")
    
    if categories:
        selected_category = st.selectbox(
            "Select Category to Edit:",
            options=[cat.get('name', 'Unknown') for cat in categories],
            key="edit_category_select"
        )
        
        if selected_category:
            category_to_edit = next((cat for cat in categories if cat.get('name') == selected_category), None)
            
            if category_to_edit:
                st.write(f"**Editing:** {category_to_edit.get('name')}")
                
                with st.form("edit_category_form"):
                    new_name = st.text_input("New Name", value=category_to_edit.get('name', ''))
                    new_description = st.text_area("New Description", value=category_to_edit.get('description', ''))
                    
                    col1, col2 = st.columns(2)
                    
                    with col1:
                        if st.form_submit_button("Update Category"):
                            result = api_client.update_category(
                                category_to_edit.get('id'),
                                new_name.strip(),
                                new_description.strip() if new_description.strip() else None
                            )
                            if result:
                                st.success("✅ Category updated successfully!")
                                st.rerun()
                            else:
                                st.error("❌ Failed to update category. Please try again.")
                    
                    with col2:
                        if st.form_submit_button("Delete Category"):
                            result = api_client.delete_category(category_to_edit.get('id'))
                            if result:
                                st.success("✅ Category deleted successfully!")
                                st.rerun()
                            else:
                                st.error("❌ Failed to delete category. Make sure it has no items.")
    else:
        st.info("No categories to edit. Create some categories first!")

# Quick navigation
st.markdown("---")
if st.button("🏠 Back to Dashboard"):
    st.switch_page("main.py")
